File: 22. Generate Parentheses. Generate Parentheses.py
def generateParenthesis(n):

    stack = []
    res = []

    def helper(left, right): 
        if n == left == right:
            res.append("".join(stack))
            return 
        if left < n:
            stack.append("(")
            helper(left + 1, right)
            stack.pop()
        if left > right:
            stack.append(")")
            helper(left, right + 1)
            stack.pop()

    helper(0, 0)
    return res


print(generateParenthesis(3))            

File: 22. Generate Parentheses. Generate Parentheses.go
package main

import "fmt"

func main() {
	fmt.Println(generateParenthesis(3))
}

func generateParenthesis(n int) []string {

	stack := make([]string, 0)
	res := make([]string, 0)

	var helper func(left int, rigth int)

	helper = func(left int, right int) {
		if left == n && right == n {
			var tmp string
			for _, v := range stack {
				tmp = tmp + v
			}
			res = append(res, tmp)
		}
		if left < n {
			stack = append(stack, "(")
			helper(left+1, right)
			stack = stack[:len(stack)-1]
		}
		if right < left {
			stack = append(stack, ")")
			helper(left, right+1)
			stack = stack[:len(stack)-1]
		}
	}

	helper(0, 0)
	return res

}
